src/octree/batch_iterator.rs

```diff
@@ -107,68 +104,74 @@ where
                 LayerData::F32(self.intensity.split_off(0)),
             );
         }
-        let point_data = PointData {
+        Some(PointData {
             position: self.position.split_off(0),
             layers,
-        };
-        (self.func)(point_data)
+        })
     }
+}
 
-    fn push_point_and_callback(&mut self, point: Point) -> Result<()> {
-        self.push_point(point);
-        if self.position.len() == self.position.capacity() {
-            return self.callback();
+impl<F> Iterator for BatchIterator<F>
+where
+    F: Fn(&Point) -> bool,
+{
+    type Item = PointData;
+    fn next(&mut self) -> Option<Self::Item> {
+        let max_points = self.position.capacity();
+        for _num_point in 0..max_points {
+            match self.point_iterator.next() {
+                Some(point) => {
+                    self.push_point(point);
+                }
+                None => return self.into_pointdata(),
+            }
         }
-        Ok(())
+        self.into_pointdata()
     }
 }
 
-/// Iterator on point batches
-pub struct BatchIterator<'a> {
-    octree: &'a Octree,
+/// compute a function while iterating on a batch of points
+pub fn try_for_each_point_batch<'a, Callback>(
+    octree: &'a octree::Octree,
     point_location: &'a PointLocation,
     batch_size: usize,
-}
-
-impl<'a> BatchIterator<'a> {
-    pub fn new(
-        octree: &'a octree::Octree,
-        point_location: &'a PointLocation,
-        batch_size: usize,
-    ) -> Self {
-        BatchIterator {
-            octree,
-            point_location,
-            batch_size,
+    mut func: Callback,
```

Comment:
this function will be splitted in the next pr:
1) gathering node ids in the octrees
2)parallelizing point iterator over nodes returning batches of points
3) applying the mut function in the receiver/consumer thread